rfcs/80-curve-trait.md

```diff
@@ -0,0 +1,793 @@
+```
+(The latter also has a convenience method, `reparametrize_by_curve`, which handles the domain automatically.)
+
+---
+
+Next, we have `graph`, the last of the general functional methods of the API:
+```rust
+/// Create a new [`Curve`] which is the graph of this one; that is, its output includes the
+/// parameter itself in the samples. For example, if this curve outputs `x` at time `t`, then
+/// the produced curve will produce `(t, x)` at time `t`.
+fn graph(self) -> impl Curve<(f32, T)>
+where
+    Self: Sized,
+{ //... }
+```
+This is a subtle method whose main applications involve allowing more complex things to be expressed. For example,
+here we modify a curve by making its output value attenuate with time:
+```rust
+// A curve with a value of `3.0` over the `Interval` from `0.0` to `5.0`:
+let my_curve = const_curve(interval(0.0, 5.0).unwrap(), 3.0);
+// The same curve but with exponential falloff in time:
+let new_curve = my_curve.graph().map(|(t, x)| x * (-t).exp2());
+```
+
+Most general operations that one could think up for curves can be achieved by clever combinations of `map` and
+`reparametrize`, perhaps with a little `graph` sprinkled in.
+
+### Sampling, resampling, and interpolation
+
+The next part of the API concerns itself with more imperative (i.e. data-focused) matters. 
+
+Often, one will want to define a curve using some kind of interpolation over discrete data or, conversely,
+extract lists of samples that approximate a curve, or convert a curve to one which has been discretized.
+
+For the first of these, there is the type `SampleCurve<T, I>`, whose constructor looks like this:
+```rust
+/// Create a new `SampleCurve`, using the specified `interpolation` to interpolate between
+/// the given `samples`. An error is returned if there are not at least 2 samples or if the
+/// given `domain` is unbounded.
+///
+/// The interpolation takes two values by reference together with a scalar parameter and
+/// produces an owned value. The expectation is that `interpolation(&x, &y, 0.0)` and
+/// `interpolation(&x, &y, 1.0)` are equivalent to `x` and `y` respectively.
+pub fn new(
+    domain: Interval,
+    samples: impl Into<Vec<T>>,
+    interpolation: I,
+) -> Result<Self, SampleCurveError>
+where
+    I: Fn(&T, &T, f32) -> T,
+{ //... }
+```
+Here, the `samples` become a `Vec<T>` whose elements represent equally-spaced sample values over the given
+`domain`. Adjacent samples are interpolated using the given `interpolation`. The result is a type that
+implements `Curve<T>`, on which the preceding operations can be performed.
+
+Conversely, if one wants discrete samples from a curve, there is the function `samples`:
+```rust
+/// Extract an iterator over evenly-spaced samples from this curve. If `samples` is less than 2
+/// or if this curve has unbounded domain, then an error is returned instead.
+fn samples(&self, samples: usize) -> Result<impl Iterator<Item = T>, ResamplingError> { //... }
```

Comment:
> This should probably take an `Into<Iterator>`.

I wrote in the document about why this isn't the case: basically, you can just `.map` over your iterator if that's what you want to do, and the behavior of requested samples that aren't within the curve domain seems like it might be sensitive to the problem at hand. The benefit of having this method as-is is that it's just really simple: for example, if I want to take my curve and get a vector of points to render into a linestrip with gizmos, I can just do something like this without giving it much thought:
```rust
gizmos.linestrip(my_curve.samples(100));
```

As for the distance function: this is a short-term goal, but I think it would be best addressed when we start trying to do geometry with curves :)